base_layer/mmr/tests/merkle_mountain_range.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+mod support;
+
+use digest::Digest;
+use support::{combine_hashes, create_mmr, int_to_hash, Hasher};
+use tari_mmr::{MerkleMountainRange, VectorBackend};
+
+/// MMRs with no elements should provide sane defaults. The merkle root must be the hash of an empty string, b"".
+#[test]
+fn zero_length_mmr() {
+    let mmr = MerkleMountainRange::<Hasher, _>::new(VectorBackend::default());
+    assert_eq!(mmr.len(), 0);
+    assert!(mmr.is_empty());
+    let empty_hash = Hasher::digest(b"").to_vec();
+    assert_eq!(&mmr.get_merkle_root(), &empty_hash);
+}
+
+/// Successively build up an MMR and check that the roots, heights and indices are all correct.
+#[test]
+fn build_mmr() {
+    let mut mmr = MerkleMountainRange::<Hasher, _>::new(VectorBackend::default());
+    // Add a single item
+    let h0 = int_to_hash(0);
+
+    assert!(mmr.push(&h0).is_ok());
+    // The root of a single hash is the hash of that hash
+    assert_eq!(mmr.len(), 1);
+    assert_eq!(mmr.get_merkle_root(), combine_hashes(&[&h0]));
+    // Two leaf item items:
+    //    2
+    //  0   1
+    let h1 = int_to_hash(1);
+    assert!(mmr.push(&h1).is_ok());
+    let h_2 = combine_hashes(&[&h0, &h1]);
+    assert_eq!(mmr.get_merkle_root(), combine_hashes(&[&h_2]));
+    assert_eq!(mmr.len(), 3);
```

Comment:
Should the len be 3 or 2?
I am of the opinion that the len should return 2 as you only added 2 items. 
Returning 3 is the correct len of the mmr, but I cannot see a use for that I only see I use for returning 2.

Reply:
I agree for a public API, but this is essentially a private class that is working with the mmr indices everywhere. The next PR build on top of this exposing a user-facing API that does exactly what you suggest